tensorboard/plugins/histogram/histograms_plugin.py

```diff
@@ -85,18 +125,74 @@ def histograms_impl(self, tag, run, downsample_to=50):
     At most `downsample_to` events will be returned. If this value is
     `None`, then no downsampling will be performed.
     """
-    try:
-      tensor_events = self._multiplexer.Tensors(run, tag)
-    except KeyError:
-      raise ValueError('No histogram tag %r for run %r' % (tag, run))
-    events = [[ev.wall_time, ev.step, tf.make_ndarray(ev.tensor_proto).tolist()]
-              for ev in tensor_events]
-    if downsample_to is not None and len(events) > downsample_to:
-      indices = sorted(random.Random(0).sample(list(range(len(events))),
-                                               downsample_to))
-      events = [events[i] for i in indices]
+    if self._db_connection_provider:
+      # Serve data from the database.
+      db = self._db_connection_provider()
+      # We select for steps greater than -1 because the writer inserts
+      # placeholder rows en masse. The check for step filters out those rows.
+      query = '''
+        SELECT
+          Tensors.computed_time AS computed_time,
+          Tensors.step AS step,
+          Tensors.data AS data,
+          Tensors.dtype AS dtype,
+          Tensors.shape AS shape
+        FROM Tensors
+        LEFT JOIN Tags
+          ON Tensors.series = Tags.tag_id
+        LEFT JOIN Runs
+          ON Tags.run_id = Runs.run_id 
+        WHERE
+          Runs.run_name = ?
+          AND Tags.tag_name = ?
+          AND Tags.plugin_name = ?
+          AND Tensors.step > -1
+      '''
+      if downsample_to is not None:
+        # Wrap the query in an outer one that samples.
+        query = '''
+          SELECT *
+          FROM 
+            (%(query)s
+             ORDER BY RANDOM()
```

Comment:
Just FYI, once we're dealing with a fixed-length series I think we can do better than random sampling (as discussed at length on the reservoir sampling thread).  This is fine for now since it's just equivalent to the non-DB behavior, but I'd like to look at changing this in a future PR (already have done some research).

Reply:
Awesome. Ack.